docs/architecture/adr-071-crypto-v2-multi-curve.md

```diff
@@ -0,0 +1,597 @@
+# ADR 071: Cryptography v2- Multi-curve support
+
+## Change log
+
+* May 7th 2024: Initial Draft (Zondax AG: @raynaudoe @juliantoledano @jleni @educlerici-zondax @lucaslopezf)
+
+## Status
+
+DRAFT
+
+## Abstract
+
+This ADR proposes the refactoring of the existing Keyring module to support multiple cryptographic curves for signing and verification processes. With this update, we aim to facilitate the integration of new cryptographic curves through clean and simple interfaces. Additionally, support for Hardware Security Modules (HSM) is introduced as a complementary enhancement in this redesign.
+
+This ADR also introduces the capability to support remote signers. This feature will enable the nodes to interact with cryptographic signers that are not locally present on the system where the main app is running. This is particularly useful for scenarios where keys are managed in secure, remote environments or when leveraging cloud-based cryptographic services.
+
+
+## Introduction
+
+The introduction of multi-curve support in the cosmos-sdk cryptographic module offers significant advantages. By not being restricted to a single cryptographic curve, developers can choose the most appropriate curve based on security, performance, and compatibility requirements. This flexibility enhances the application's ability to adapt to evolving security standards and optimizes performance for specific use cases, helping to future-proofing the sdk's cryptographic capabilities.
```

Comment:
Sdk already supports multiple curves. Ethan added zk as signatures as well. I don't think this is the correct phrasing. I think the current method is simpler than what is being proposed here.

Reply:
I understand your point, but in the previous round of reviews a few months ago we've agreed that that wasn't 'that easy' to do as it might be now so the need for some new strategy like the one we propose was needed. 
Again, that was almost 6 months ago I know that the sdk's code have evolved from that time and maybe there are 'now' some implementations that might help to achieve what we mention in the ADR.

Putting curves aside, the sdk is still not capable of managing HSM and other kind of remote keys providers in a modular fashion. Since the solution proposed works for those cases, adding new curves on the other hand is almost 'free' in terms of the framework needed in the sdk to have that kind of modularity.

Reply:
I'm afraid I strongly disagree here, just search in the sdk code the string 'ledger' and you'll see how many direct references and conditions are affected by this particular signing method. Just thinking of adding for example another kind of hw wallet, means duplicating code paths for that particular device and having in the code things like `ifI(IsLedgerDevice)` `if(IsOneKey)`, `If(IsAnotherHW)`, etc. The same reasoning, at different degrees of complexity, can be extrapolated for when adding other curves or cloud based HSM. 
The provider idea is to abstract all of these things into one single package, I like to think of it as a library that has different backends to perform the same operations.